rfcs/20240808-graph-api-int-compression-for-sdpa/README.md

```diff
@@ -0,0 +1,328 @@
+      inputs should be `1d` tensors.
+   2. For `per_group` quantization:
+      - On the dimension where the group quantization is applied, the dimension
+      should equal to `src_dim / group_size`.
+      - On other dimensions, the dimension should match the input.
+
+For example, if the shape of Key tensor is `4x8x4`, and the group shape is
+`1x4x1`, which means each scaling factor will be adopted for 4 times. One
+possible scale would be:
+
+![alt text](img/grouped_scale.png)
+
+The usage of the new attribute will be like:
+
+```cpp
+using namespace dnnl::graph;
+
+const dim K = 10, N = 20;
+const size_t nG = 2, G = N / nG;
+
+dims src_dims = {K, N};
+dims scale_dims = {K, N/G};
+dims group_shape = {1, G}
+
+// In real case, users do not have to provide the concrete shapes of logical
+// tensors until compilation stage. 
+logical_tensor src_desc {0, data_type::u4, src_dims, layout_type::strided};
+logical_tensor scale_desc {
+        1, data_type::f32, scale_dims, layout_type::strided};
+logical_tensor dst_desc {2, data_type::f32, src_dims, layout_type::strided};
+
+op deq = op(0, op::kind::DynamicDequantize, {src_desc, scale_desc},
+        {dst_desc}, "deq");
+// Specify the quantization type as per_group quantization.
+deq.set_attr<std::string>(op::attr::qtype, "per_group");
+// Axis indicates on which dimension the quantization will be applied.
+deq.set_attr<int64_t>(op::attr::axis, 1);
+// Group size indicates the the size of each group.
+deq.set_attr<int64_t>(op::attr::group_shape, group_shape);
+
+// Create graph and add the op to the graph
+graph g(dnnl::engine::kind::gpu);
+g.add_op(deq);
+
+// ...
+```
+
+##### Limitation
+
+With the introduction of the new attribute `group_shape`, Graph API now
+enables users to define group quantization across multiple dimensions. This
+design aims to accommodate potential quantization types, such as per-token
+quantization, where the group size is defined as H*D. However, these
+quantization types are not yet supported by the current implementation. As for
+grouped quantization, the backend currently facilitates quantization on one of
+the last two dimensions, while all other inputs will be deemed invalid.
+
+#### Alternative options
+
+Similar to the discussion of int8 and fp8 data types, an alternative solution
+may be supporting int4 data types directly in computation operations like MatMul
+and Convolution. But it will bloat and complicate the opset and op schema of
+oneDNN Graph. Hence it's not considered here.
+
+In addition, to enhance clarity and reduce compilation overhead, oneDNN Graph
+will support the direct quantization between int4 data types and bf16/f16.
+This means that additional `Typecast` operations will no longer be necessary
+between quantization processes and subsequent operations in bf16/f16 computation
+graphs. Although it will complicate the fusion patterns to some extend, we can
+address this by implementing more precise pattern definitions.
```

Comment:
It seems to me that this paragraph belongs to a `Proposal 3` section instead since it already has TypeCast dropped from the picture.